cpp/arcticdb/util/std_ranges_utils.hpp

```diff
@@ -0,0 +1,165 @@
+#pragma once
+
+#include <ranges>
+#include <unordered_map>
+#include <vector>
+#include <algorithm>
+#include <functional>
+#include <optional>
+#include <stack>
+#include <memory>
+
+namespace arcticdb::util {
+
+// Standard library replacement for folly::gen::groupBy
+template<typename Container, typename KeyFunc>
+auto group_by(Container&& container, KeyFunc&& key_func) {
+    // Use decltype to get the correct value type from the range
+    using ValueType = std::remove_reference_t<decltype(*std::begin(container))>;
+    using KeyType = std::invoke_result_t<KeyFunc, ValueType>;
+    
+    std::unordered_map<KeyType, std::vector<ValueType>> result;
+    
+    for (auto&& item : std::forward<Container>(container)) {
+        auto key = std::invoke(std::forward<KeyFunc>(key_func), item);
+        result[std::move(key)].push_back(std::forward<ValueType>(item));
+    }
+    
+    return result;
+}
+
+// Standard library replacement for folly::gen::zip with move semantics
+template<typename Container1, typename Container2>
+auto zip_move(Container1&& container1, Container2&& container2) {
+    using ValueType1 = std::remove_reference_t<decltype(*std::begin(container1))>;
+    using ValueType2 = std::remove_reference_t<decltype(*std::begin(container2))>;
+    
+    std::vector<std::pair<ValueType1, ValueType2>> result;
+    result.reserve(std::min(container1.size(), container2.size()));
+    
+    auto it1 = std::begin(container1);
+    auto it2 = std::begin(container2);
+    auto end1 = std::end(container1);
+    auto end2 = std::end(container2);
+    
+    while (it1 != end1 && it2 != end2) {
+        result.emplace_back(std::move(*it1), std::move(*it2));
+        ++it1;
+        ++it2;
+    }
+    
+    return result;
+}
+
+// Helper for foreach on grouped data
+template<typename GroupedContainer, typename Func>
+void foreach_group(GroupedContainer&& grouped, Func&& func) {
```

Comment:
Same as in group_by regarding moving func and using move iterators on the container